lib/Parse/ParseExpr.cpp

```diff
@@ -2905,9 +2918,101 @@ ExprResult Parser::ParseBoundsExpression() {
     SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
 
   PT.consumeClose();
+
   return Result;
 }
 
+void Parser::SkipInvalidExprAndParseRelativeBoundsClause(ExprResult &Result,
+                                                         bool isBoundsorItype) {
+  Token Next;
+  if (!isBoundsorItype) {
+    Diag(Tok, diag::err_expected_bounds_expr_or_interop_type);
+    while (1) {
+      Next = NextToken();
+      BalancedDelimiterTracker PT(*this, tok::l_paren);
+      if (Tok.getKind() == tok::l_paren) {
+        PT.consumeOpen();
+        SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
+        PT.consumeClose();
+      }
+
+      if (StartsRelativeBoundsClause(Tok)) {
+        ParseRelativeBoundsClause(Result);
+        ConsumeAnyToken();
+        break;
+      }
+
+      if (Tok.isOneOf(tok::equal, tok::r_paren, tok::semi, tok::comma,
+                      tok::l_brace)) {
+        break;
+      }
+      ConsumeAnyToken();
+    }
+  } else {
+    if (StartsRelativeBoundsClause(Tok)) {
+      ParseRelativeBoundsClause(Result);
+    }
+  }
+}
+
+bool Parser::ParseRelativeBoundsClause(ExprResult &Expr) {
+  bool IsError = false;
+  RelativeBoundsClause *RelativeClause = nullptr;
+  ExprResult ConstExpr;
+  RangeBoundsExpr *Range = nullptr;
+
+  IdentifierInfo *Ident = Tok.getIdentifierInfo();
+  SourceLocation BoundsKWLoc = Tok.getLocation();
+  ConsumeToken();
+
+  BalancedDelimiterTracker PT(*this, tok::l_paren);
+
+  if (PT.expectAndConsume(diag::err_expected_lparen_after,
+                          Ident->getNameStart())) {
+    SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
+    IsError = true;
+    return IsError;
+  }
+
+  if (Ident == Ident_rel_align) {
+    TypeResult ResultTy = ParseTypeName();
+    if (ResultTy.isInvalid()) {
+      SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
+      IsError = true;
+    } else
+      RelativeClause = Actions.ActOnRelativeTypeBoundsClause(
+          BoundsKWLoc, ResultTy.get(), Tok.getLocation());
+  } else if (Ident == Ident_rel_align_value) {
+    ConstExpr = ParseConstantExpression();
+    if (ConstExpr.isInvalid()) {
+      SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
+      IsError = true;
+    } else {
+      ConstExpr = Actions.VerifyIntegerConstantExpression(ConstExpr.get());
+      if (!ConstExpr.isInvalid())
+        RelativeClause = Actions.ActOnRelativeConstExprClause(
+            ConstExpr.get(), BoundsKWLoc, Tok.getLocation());
+    }
+  } else {
+    llvm_unreachable("unexpected relative alignment clause kind");
+  }
+
+  if (Expr.isInvalid()) {
+    SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
```

Comment:
I think you should just return 'true' here. We don't know why the Expr that was passed in was invalid, so we should not emit an error message. The programmer will need to fix the error that caused the invalid expression first.